File: CS/E2371.Module/SetupAuthenticationHelper.cs
using DevExpress.ExpressApp;
using DevExpress.ExpressApp.Security;
using DevExpress.Persistent.BaseImpl.PermissionPolicy;
using E2371.Module.BusinessObjects;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace E2371.Module {
    public static class SetupAuthenticationHelper {
        public static void Setup(XafApplication application) {
            AuthenticationMixed mixedAuthentication = new AuthenticationMixed();
            mixedAuthentication.LogonParametersType = typeof(MixedLogonParameters);
            mixedAuthentication.IsSupportChangePassword = true;
            CustomAuthenticationStandardProvider authenticationStandardProvider = new CustomAuthenticationStandardProvider(application.Security.UserType);
            AuthenticationActiveDirectoryProvider authenticationActiveDirectoryProvider = new AuthenticationActiveDirectoryProvider(application.Security.UserType, (ICanInitializeNewUser)application.Security, true);
            authenticationActiveDirectoryProvider.CustomCreateUser += AuthenticationActiveDirectoryProvider_CustomCreateUser;
            mixedAuthentication.AuthenticationProviders.Add(typeof(AuthenticationStandardProvider).Name, authenticationStandardProvider);
            mixedAuthentication.AuthenticationProviders.Add(typeof(AuthenticationActiveDirectoryProvider).Name, authenticationActiveDirectoryProvider);
            ((SecurityStrategyComplex)application.Security).Authentication = mixedAuthentication;
            application.LoggingOn += SecurityMixedAuthenticationWindowsFormsApplication_LoggingOn;
        }
        private static void SecurityMixedAuthenticationWindowsFormsApplication_LoggingOn(object sender, LogonEventArgs e) {
            XafApplication application = (XafApplication)sender;
            AuthenticationMixed authenticationMixed = (AuthenticationMixed)((SecurityStrategyComplex)application.Security).Authentication;
            switch(((MixedLogonParameters)e.LogonParameters).AuthenticationType) {
                case AuthenticationType.Application:
                    authenticationMixed.SetupAuthenticationProvider(typeof(AuthenticationStandardProvider).Name, e.LogonParameters);
                    break;
                case AuthenticationType.Windows:
                    authenticationMixed.SetupAuthenticationProvider(typeof(AuthenticationActiveDirectoryProvider).Name);
                    break;
            }
        }
        private static void AuthenticationActiveDirectoryProvider_CustomCreateUser(object sender, CustomCreateUserEventArgs e) {
            AuthenticationActiveDirectoryProvider provider = (AuthenticationActiveDirectoryProvider)sender;
            CustomPermissionPolicyUser user = e.ObjectSpace.CreateObject<CustomPermissionPolicyUser>();
            user.UserName = e.UserName;
            user.IsActiveDirectoryUser = true;
            provider.CanInitializeNewUser.InitializeNewUser(e.ObjectSpace, user);
            e.User = user;
            e.Handled = true;
        }
    }
}

File: VB/E2371.Module/SetupAuthenticationHelper.vb
Imports Microsoft.VisualBasic
Imports DevExpress.ExpressApp
Imports DevExpress.ExpressApp.Security
Imports DevExpress.Persistent.BaseImpl.PermissionPolicy
Imports E2371.Module.BusinessObjects
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text
Imports System.Threading.Tasks

Namespace E2371.Module
	Public NotInheritable Class SetupAuthenticationHelper
		Private Sub New()
		End Sub
		Public Shared Sub Setup(ByVal application As XafApplication)
			Dim mixedAuthentication As New AuthenticationMixed()
			mixedAuthentication.LogonParametersType = GetType(MixedLogonParameters)
			mixedAuthentication.IsSupportChangePassword = True
			Dim authenticationStandardProvider As New CustomAuthenticationStandardProvider(application.Security.UserType)
			Dim authenticationActiveDirectoryProvider As New AuthenticationActiveDirectoryProvider(application.Security.UserType, CType(application.Security, ICanInitializeNewUser), True)
			AddHandler authenticationActiveDirectoryProvider.CustomCreateUser, AddressOf AuthenticationActiveDirectoryProvider_CustomCreateUser
			mixedAuthentication.AuthenticationProviders.Add(GetType(AuthenticationStandardProvider).Name, authenticationStandardProvider)
			mixedAuthentication.AuthenticationProviders.Add(GetType(AuthenticationActiveDirectoryProvider).Name, authenticationActiveDirectoryProvider)
			CType(application.Security, SecurityStrategyComplex).Authentication = mixedAuthentication
			AddHandler application.LoggingOn, AddressOf SecurityMixedAuthenticationWindowsFormsApplication_LoggingOn
		End Sub
		Private Shared Sub SecurityMixedAuthenticationWindowsFormsApplication_LoggingOn(ByVal sender As Object, ByVal e As LogonEventArgs)
			Dim application As XafApplication = CType(sender, XafApplication)
			Dim authenticationMixed As AuthenticationMixed = CType((CType(application.Security, SecurityStrategyComplex)).Authentication, AuthenticationMixed)
			Select Case (CType(e.LogonParameters, MixedLogonParameters)).AuthenticationType
				Case AuthenticationType.Application
					authenticationMixed.SetupAuthenticationProvider(GetType(AuthenticationStandardProvider).Name, e.LogonParameters)
				Case AuthenticationType.Windows
					authenticationMixed.SetupAuthenticationProvider(GetType(AuthenticationActiveDirectoryProvider).Name)
			End Select
		End Sub
		Private Shared Sub AuthenticationActiveDirectoryProvider_CustomCreateUser(ByVal sender As Object, ByVal e As CustomCreateUserEventArgs)
			Dim provider As AuthenticationActiveDirectoryProvider = CType(sender, AuthenticationActiveDirectoryProvider)
			Dim user As CustomPermissionPolicyUser = e.ObjectSpace.CreateObject(Of CustomPermissionPolicyUser)()
			user.UserName = e.UserName
			user.IsActiveDirectoryUser = True
			provider.CanInitializeNewUser.InitializeNewUser(e.ObjectSpace, user)
			e.User = user
			e.Handled = True
		End Sub
	End Class
End Namespace
